Change threads to tree-like view in web UI

mumot1999 opened the pull request:
Hello, I have created feature for simplifying view of responses for statuses.

Before: 
![Screenshot from 2022-01-12 11-27-09](https://user-images.githubusercontent.com/24902533/149123585-b915b10b-a96d-4890-844c-f2378514155d.png)

After:
![Screenshot from 2022-01-12 11-27-49](https://user-images.githubusercontent.com/24902533/149123582-fbdaeaad-7243-4290-a433-a6c655fe86c4.png)

mumot1999 committed:
feat: add 1 level indents for replies in thread

mumot1999 committed:
feat: add custom level deepness of responses

mumot1999 committed:
feat: add setResponsesWithLinesData function

mumot1999 committed:
feat: add lines on posts replies

mumot1999 committed:
refactor: reformat scss file

mumot1999 commented:
Any tool for auto resolve codeclimate?

ClearlyClaire commented:
Hello, thank you for your pull request!

I have not reviewed the code at all yet, so I'm only commenting based on the screenshots: this looks surprisingly good, and makes understanding forking threads easier. However, I am not completely sure how well it would work in practice with long threads. Furthermore, a very common usage pattern (maybe a bit less on Mastodon than Twitter thanks to the increased characters limit) is to post self-replies in a linear thread, which in this case would increase the depth on each post.

For these reasons I'm not sure this should be the only—or even default—way of rendering threads. But it is definitely worth investigating!

> Any tool for auto resolve codeclimate?

As far as I know, the CodeClimate warnings come from `eslint`. You could try `eslint --fix`, although I have never tried it myself.

Note also that not 100% of our existing code adheres to the coding styles checked by CodeClimate, so using `eslint --fix` is likely to produce changes in other parts of the code, and that should not be submitted as part of this PR.

mumot1999 commented on app/javascript/mastodon/features/status/index.js:
```diff
@@ -514,8 +549,22 @@ class Status extends ImmutablePureComponent {
       ancestors = <div>{this.renderChildren(ancestorsIds)}</div>;
     }
 
+    const wrapDescendants = (elements) => {
```
sorry for force push, missed I have added wrong file

mumot1999 commented:
Sorry for force push, missed I have added wrong file.

> I am not completely sure how well it would work in practice with long threads. 

I have added parameter for setting maximum tree depth. Currently it's hardcoded for 3.

What do you think about place a button in right corner for changing the view of threads. 
User could choose tree view or list.

ClearlyClaire commented:
I'm trying it out on a real instance to get a sense of whether the design works well or not, and apparently, this does not work well on multi-column or mobile (single-column on a wide screen does not have this issue):
![image](https://user-images.githubusercontent.com/384364/149300525-ed51a1be-a348-439c-9784-a568537659cf.png)

> What do you think about place a button in right corner for changing the view of threads.
> User could choose tree view or list.

That sounds like a good idea!
